src/reporter/mod.rs

```diff
@@ -0,0 +1,284 @@
+use crate::commands::DEFAULT_CONFIG_PATH;
+use crate::settings::{self, toml::Manifest};
+
+use std::env::args;
+use std::fs::{read_dir, File};
+use std::io::{BufReader, Write};
+use std::panic;
+use std::panic::PanicInfo;
+use std::path::{Path, PathBuf};
+use std::{collections::HashMap, fs};
+
+use anyhow::{anyhow, Result};
+use backtrace::{Backtrace, BacktraceFrame};
+use path_slash::PathExt;
+use serde::{Deserialize, Serialize};
+use sys_info::{os_release, os_type};
+use uuid::Uuid;
+
+const PANIC_UNWIND_START_MARKER: &str = "rust_begin_unwind";
+const BACKTRACE_PATH_PREFIX: &str = "backtrace::";
+
+#[derive(Clone, Debug, Deserialize, Serialize)]
+pub struct Report {
+    uuid: Uuid,
+    timestamp_ms: u128,
+    host_env: HashMap<String, String>, // "os": "..." TODO: consider struct over HashMaps
+    project_info: HashMap<String, String>,
+    args: Vec<String>,
+    panic: Option<String>,
+    location: Option<String>,
+    backtrace: String,
+}
+
+/// Overrides any panic hooks with wrangler's error reporting, which logs error reports to disl with
+/// details from a panic and useful information from the wrangler user's system for debugging.
+pub fn init() {
+    // TODO: consider using panic::take_hook, and showing the original panic to the end-user without
+    // polluting the console to the point the wrangler error report message is lost in the noise.
+    panic::set_hook(Box::new(|panic_info| generate_report(Some(panic_info))));
+}
+
+/// finds the most recent log in the error report directory, unless `log` is Some, and will try to
+/// read that file instead.
+pub fn read_log(log: Option<&str>) -> Result<Report> {
+    match log {
+        Some(path) => {
+            let r = BufReader::new(File::open(error_report_dir()?.join(path))?);
+            serde_json::from_reader(r).map_err(|e| e.into())
+        }
+        None => latest_report(),
+    }
+}
+
+/// gathers necessary error report information, and stores on disk until uploaded. the
+pub fn generate_report(panic_info: Option<&PanicInfo>) {
+    let mut report = Report {
+        uuid: Uuid::new_v4(),
+        timestamp_ms: 0,
+        host_env: load_host_info(),
+        project_info: load_project_info(),
+        args: args().collect::<Vec<_>>(),
+        panic: panic_payload(panic_info),
+        location: None,
+        backtrace: useful_frames(),
+    };
+
+    if let Some(info) = panic_info {
+        if let Some(loc) = info.location() {
+            report.location = Some(format!(
+                "{}, line: {} column: {}",
+                loc.file(),
+                loc.line(),
+                loc.column()
+            ));
+        }
+    }
+
+    match std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH) {
+        Ok(ts) => {
+            report.timestamp_ms = ts.as_millis();
+        }
+        Err(e) => {
+            err_exit(format!("system time failure: {}", e), 1);
```

Comment:
it's _really_ unlikely that this will happen. if it does happen there's something else really wrong with the users system. 

using a value like `-1` would be a little awkward to detect, because it will always be at the opposite end of the sorted list from where we're looking (we want the biggest number == most recent created). the other issue is that as a single value, the file contents will be overwritten if another panic occurs and we'd lose the previous data. 

let's hold off on this for now and see if anyone has issues.